### PR TITLE
Initial JupyterLab Notebook Support

### PR DESCRIPTION
**Pull request by anantmittal**

### CODE REVIEW

**Comment by anantmittal** [comment]
@jayfo Incorporated the discussed change in `Sample Notebook.py`

**Review by jayfo — APPROVED** [review]

**Comment by jayfo** [comment]
@anantmittal : Can you now review my changes?

- The save hook should always erase all output from a notebook before writing it to disk.
- We must remember these notebooks are in a public repository, they must not include identifying information. I included a sample notebook that references specific participants by document ID, because that's literally a random string.
- We must remember to preview commits to inspect that the save hook has worked, ensure there's no output in the notebook. If that somehow gets committed at some point, it's going to be a lot of work to remove it from the git history.
- Decrypting the archive currently takes minutes. That's annoying but "fine" because all later cells then already have it loaded.

**Comment by jayfo** [comment]
@anantmittal : Are you getting this error on each startup?

Capture

I currently am. Not sure if it's an issue in our setup, or something went bad on my local machine.

File: notebooks/Pipfile [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+jupyter-dash = "*"
+numpy = "*"
+pandas = "*"
+
+jupyterlab-code-formatter = "*"
+black = "*"
+isort = "*"
+
+jupytext = "*"
```
[the remark this answers is not in the document]
Comment (anantmittal) [reply notebooks/Pipfile:20]: Error on `invoke depend.install.notebooks`:
image

[the remark this answers is not in the document]
Comment (anantmittal) [reply notebooks/Pipfile:20]: Error on `invoke depend.update.notebooks`:

**Comment by anantmittal** [comment]
Error on `invoke depend.install.notebooks`:
image

**Comment by anantmittal** [comment]
Error on `invoke depend.update.notebooks`:
image

**Comment by anantmittal** [comment]
> @anantmittal : Are you getting this error on each startup?
> 
> Capture
> 
> I currently am. Not sure if it's an issue in our setup, or something went bad on my local machine.

Yeah, I am getting this error. The jupyter kernel also seems to keep crashing.

**Comment by jayfo** [comment]
> Error on `invoke depend.install.notebooks`: image

I've never managed to fix this kind of an error when it happens. What's happening is that it's trying to uninstall a package that's installed. But for some reason it's failing at that, hence the `umpy` instead of `numpy`. All I know is to blow up the virtual environment entirely.

> pipenv_activate.bat
> cd notebooks
> pipenv --rm
> cd ..
> invoke depend.install.notebooks

Can that get an install to succeed?

**Comment by jayfo** [comment]
> Error on `invoke depend.update.notebooks`: image

You shouldn't need to update.

**Comment by anantmittal** [comment]
> > Error on `invoke depend.install.notebooks`: image
> 
> I've never managed to fix this kind of an error when it happens. What's happening is that it's trying to uninstall a package that's installed. But for some reason it's failing at that, hence the `umpy` instead of `numpy`. All I know is to blow up the virtual environment entirely.
> 
> > pipenv_activate.bat
> > cd notebooks
> > pipenv --rm
> > cd ..
> > invoke depend.install.notebooks
> 
> Can that get an install to succeed?

Yep, this worked.

**Comment by anantmittal** [comment]
The jupyter kernel keeps crashing. Investigating the issue. 

Snippet from console

>   File "/Users/a42/.local/share/virtualenvs/notebooks-4CjZOhb5/lib/python3.9/site-packages/IPython/utils/_process_posix.py", line 23, in <module>
    import pexpect
ModuleNotFoundError: No module named 'pexpect'
[W 2023-07-14 10:55:06.671 ServerApp] AsyncIOLoopKernelRestarter: restart failed
[W 2023-07-14 10:55:06.672 ServerApp] Kernel 3f20d555-fc8e-47b5-af36-ea45113697d2 died, removing from map.
[W 2023-07-14 10:55:08.486 ServerApp] 404 GET /api/kernels/3f20d555-fc8e-47b5-af36-ea45113697d2?1689357308480 (127.0.0.1): Kernel does not exist: 3f20d555-fc8e-47b5-af36-ea45113697d2
[W 2023-07-14 10:55:08.486 ServerApp] wrote error: 'Kernel does not exist: 3f20d555-fc8e-47b5-af36-ea45113697d2'
    Traceback (most recent call last):
      File "/Users/a42/.local/share/virtualenvs/notebooks-4CjZOhb5/lib/python3.9/site-packages/tornado/web.py", line 1786, in _execute
        result = await result
      File "/Users/a42/.local/share/virtualenvs/notebooks-4CjZOhb5/lib/python3.9/site-packages/jupyter_server/services/kernels/handlers.py", line 73, in get
        model = await ensure_async(km.kernel_model(kernel_id))
      File "/Users/a42/.local/share/virtualenvs/notebooks-4CjZOhb5/lib/python3.9/site-packages/jupyter_server/services/kernels/kernelmanager.py", line 501, in kernel_model
        self._check_kernel_id(kernel_id)
      File "/Users/a42/.local/share/virtualenvs/notebooks-4CjZOhb5/lib/python3.9/site-packages/jupyter_server/services/kernels/kernelmanager.py", line 532, in _check_kernel_id
        raise web.HTTPError(404, "Kernel does not exist: %s" % kernel_id)
    tornado.web.HTTPError: HTTP 404: Not Found (Kernel does not exist: 3f20d555-fc8e-47b5-af36-ea45113697d2)
[W 2023-07-14 10:55:08.487 ServerApp] 404 GET /api/kernels/3f20d555-fc8e-47b5-af36-ea45113697d2?1689357308480 (4c79aeb9b70e4f8eac8d4b4bafd2f00d@127.0.0.1) 3.36ms referer=http://localhost:8888/lab/tree/issues/issue442/issue442.ipynb
[W 2023-07-14 10:55:14.602 ServerApp] Timeout waiting for kernel_info reply from a45834a1-3b54-4c84-85c7-ebda37714e60
[I 2023-07-14 10:55:14.607 ServerApp] Connecting to kernel a45834a1-3b54-4c84-85c7-ebda37714e60.
[W 2023-07-14 10:55:14.611 ServerApp] 404 GET /api/kernels/a45834a1-3b54-4c84-85c7-ebda37714e60/channels?session_id=f999a3c6-cea6-4386-b825-e42efb7a63f8 (127.0.0.1): Kernel does not exist: a45834a1-3b54-4c84-85c7-ebda37714e60
[I 2023-07-14 10:55:16.493 ServerApp] Saving file at /issues/issue442/issue442.ipynb
[I 2023-07-14 10:55:16.496 ServerApp] Saving issue442.ipynb
[W 2023-07-14 10:55:48.482 ServerApp] Timeout waiting for kernel_info reply from 3f20d555-fc8e-47b5-af36-ea45113697d2
[I 2023-07-14 10:55:48.486 ServerApp] Connecting to kernel 3f20d555-fc8e-47b5-af36-ea45113697d2.
[W 2023-07-14 10:55:48.490 ServerApp] 404 GET /api/kernels/3f20d555-fc8e-47b5-af36-ea45113697d2/channels?session_id=f100498e-a325-4a96-99b2-b703506962b1 (127.0.0.1): Kernel does not exist: 3f20d555-fc8e-47b5-af36-ea45113697d2

**Comment by anantmittal** [comment]
> The jupyter kernel keeps crashing. Investigating the issue.
> 
> Snippet from console
> 
> > File "/Users/a42/.local/share/virtualenvs/notebooks-4CjZOhb5/lib/python3.9/site-packages/IPython/utils/_process_posix.py", line 23, in 
> > import pexpect
> > ModuleNotFoundError: No module named 'pexpect'
> > [W 2023-07-14 10:55:06.671 ServerApp] AsyncIOLoopKernelRestarter: restart failed
> > [W 2023-07-14 10:55:06.672 ServerApp] Kernel 3f20d555-fc8e-47b5-af36-ea45113697d2 died, removing from map.
> > [W 2023-07-14 10:55:08.486 ServerApp] 404 GET /api/kernels/3f20d555-fc8e-47b5-af36-ea45113697d2?1689357308480 (127.0.0.1): Kernel does not exist: 3f20d555-fc8e-47b5-af36-ea45113697d2
> > [W 2023-07-14 10:55:08.486 ServerApp] wrote error: 'Kernel does not exist: 3f20d555-fc8e-47b5-af36-ea45113697d2'
> > Traceback (most recent call last):
> > File "/Users/a42/.local/share/virtualenvs/notebooks-4CjZOhb5/lib/python3.9/site-packages/tornado/web.py", line 1786, in _execute
> > result = await result
> > File "/Users/a42/.local/share/virtualenvs/notebooks-4CjZOhb5/lib/python3.9/site-packages/jupyter_server/services/kernels/handlers.py", line 73, in get
> > model = await ensure_async(km.kernel_model(kernel_id))
> > File "/Users/a42/.local/share/virtualenvs/notebooks-4CjZOhb5/lib/python3.9/site-packages/jupyter_server/services/kernels/kernelmanager.py", line 501, in kernel_model
> > self._check_kernel_id(kernel_id)
> > File "/Users/a42/.local/share/virtualenvs/notebooks-4CjZOhb5/lib/python3.9/site-packages/jupyter_server/services/kernels/kernelmanager.py", line 532, in _check_kernel_id
> > raise web.HTTPError(404, "Kernel does not exist: %s" % kernel_id)
> > tornado.web.HTTPError: HTTP 404: Not Found (Kernel does not exist: 3f20d555-fc8e-47b5-af36-ea45113697d2)
> > [W 2023-07-14 10:55:08.487 ServerApp] 404 GET /api/kernels/3f20d555-fc8e-47b5-af36-ea45113697d2?1689357308480 (4c79aeb9b70e4f8eac8d4b4bafd2f00d@127.0.0.1) 3.36ms referer=http://localhost:8888/lab/tree/issues/issue442/issue442.ipynb
> > [W 2023-07-14 10:55:14.602 ServerApp] Timeout waiting for kernel_info reply from a45834a1-3b54-4c84-85c7-ebda37714e60
> > [I 2023-07-14 10:55:14.607 ServerApp] Connecting to kernel a45834a1-3b54-4c84-85c7-ebda37714e60.
> > [W 2023-07-14 10:55:14.611 ServerApp] 404 GET /api/kernels/a45834a1-3b54-4c84-85c7-ebda37714e60/channels?session_id=f999a3c6-cea6-4386-b825-e42efb7a63f8 (127.0.0.1): Kernel does not exist: a45834a1-3b54-4c84-85c7-ebda37714e60
> > [I 2023-07-14 10:55:16.493 ServerApp] Saving file at /issues/issue442/issue442.ipynb
> > [I 2023-07-14 10:55:16.496 ServerApp] Saving issue442.ipynb
> > [W 2023-07-14 10:55:48.482 ServerApp] Timeout waiting for kernel_info reply from 3f20d555-fc8e-47b5-af36-ea45113697d2
> > [I 2023-07-14 10:55:48.486 ServerApp] Connecting to kernel 3f20d555-fc8e-47b5-af36-ea45113697d2.
> > [W 2023-07-14 10:55:48.490 ServerApp] 404 GET /api/kernels/3f20d555-fc8e-47b5-af36-ea45113697d2/channels?session_id=f100498e-a325-4a96-99b2-b703506962b1 (127.0.0.1): Kernel does not exist: 3f20d555-fc8e-47b5-af36-ea45113697d2

Fix on mac: `pip install pexpect` followed by `pip install appnope`

**Comment by anantmittal** [comment]
> @anantmittal : Can you now review my changes?
> 
>     * The save hook should always erase all output from a notebook before writing it to disk.
**This is working.** 
> 
>     * We must remember these notebooks are in a public repository, they must not include identifying information. I included a sample notebook that references specific participants by document ID, because that's literally a random string.
> 
>     * We must remember to preview commits to inspect that the save hook has worked, ensure there's no output in the notebook. If that somehow gets committed at some point, it's going to be a lot of work to remove it from the git history.
> 
>     * Decrypting the archive currently takes minutes. That's annoying but "fine" because all later cells then already have it loaded.

**Comment by jayfo** [comment]
I don't think we strictly need `jupytext`. I had removed it as part of trying to debug the `dash` error, then put it back when I established that wasn't the issue. I'm fine keeping it around for now in case some future step for our notebooks does want to include that. I still think full text notebooks would be way easier to see / review.

**Review by jayfo — APPROVED** [review]

**Pull request merged**